sql/driver/driver_test.go

```diff
@@ -45,14 +54,74 @@ func cleanup(s *server.TestServer, db *sql.DB) {
 	s.Stop()
 }
 
+func TestDates(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	// From https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+	locationNames := []string{
+		"Canada/Newfoundland", // half-hour zone
+		"Europe/London",       // same as UTC, except for DST
+		"Pacific/Kiritimati",  // maximum positive offset
+		"Pacific/Midway",      // maximum negative offset
+		"US/Pacific",          // negative offset
+		"UTC",
+	}
+
+	for _, locationName := range locationNames {
+		loc, err := time.LoadLocation(locationName)
+		if err != nil {
+			t.Error(err)
+			continue
+		}
+
+		s, db := setup(t, loc)
+		defer cleanup(s, db)
+
+		var timestamp *time.Time
+		var date *driver.Date
+
+		for _, year := range []int{
+			1200, // distant past
+			2020, // present day, for DST rules
+			4000, // distant future
+		} {
+			for _, month := range []time.Month{
+				time.December, // winter
+				time.August,   // summer
+			} {
+				for hour := 0; hour < 24; hour++ {
+					if err := db.QueryRow(
+						"SELECT $1, $1::DATE",
+						time.Date(year, month, 20, hour, 34, 45, 123, loc),
```

Comment:
Save this value and use it as `expected` below, instead of just comparing that two different values from the database are consistent with each other.